File: euler/python/P187.py
"""
Semiprimes
Problem 187
"""

UPPER_LIMIT = 1000000
LOWEST_PRIME = 2

if __name__ == '__main__':
    # We generate prime numbers using the sieve of eratosthenes
    count, erato, primes = 0, [True] * UPPER_LIMIT, []

    # A function to find the index of the prime not exceeding the given value. Returns the index of
    # the element plus one. This is equivalent to the primes that can be multiplied.
    def prime_index(maximum, start, end):
        mid_point = (end + start) / 2
        if primes[end] <= maximum:
            return end + 1
        elif end - start <= 1:
            return start + 1
        elif primes[mid_point] > maximum:
            return prime_index(maximum, start, mid_point)
        else:
            return prime_index(maximum, mid_point, end)

    # We only need to check up to UPPER_LIMIT / 2 since we are multiplying the prime
    for prime in xrange(LOWEST_PRIME, UPPER_LIMIT / 2):
        if erato[prime]:
            primes.append(prime)
            count += prime_index(UPPER_LIMIT / prime, 0, len(primes) - 1)
            for composite in xrange(prime * prime, UPPER_LIMIT, prime):
                erato[composite] = False

    # Print out the final result
    print count

File: euler/c++/P187.cc
/*
  Semiprimes
  Problem 187
*/

#include <vector>
#include <iostream>

const uint64_t UPPER_LIMIT = 10000000;
const uint64_t LOWEST_PRIME = 2;
using namespace std;

// Returns the index of the prime number not exceeding the given maximum.
uint64_t prime_index(vector <uint64_t> * primes, uint64_t maximum, uint64_t start, uint64_t end) {
  uint64_t mid_point = (start + end) / 2;
  if((*primes)[end] <= maximum) return end + 1;
  else if(end - start <= 1) return start + 1;
  else if((*primes)[mid_point] > maximum) return prime_index(primes, maximum, start, mid_point);
  else return prime_index(primes, maximum, mid_point, end);
}

int main(int argc, char * argv[]) {
  // We generate a list of prime numbers using the sieve of Eratosthenes
  uint64_t total_sum = 0;
  vector <uint64_t> primes;
  bool * erato = new bool[UPPER_LIMIT]();

  // We only need to check up to UPPER_LIMIT / 2 as a maximum
  for(uint64_t i = LOWEST_PRIME; i < UPPER_LIMIT / 2; i++) {
    if(erato[i]) continue;
    primes.push_back(i);
    total_sum += prime_index(&primes, UPPER_LIMIT / i, 0, primes.size() - 1);
    for(uint64_t j = i * i; j < UPPER_LIMIT; j += i) erato[j] = true;
  }

  // Print out the final result and collect garbage
  cout << total_sum << endl;
  delete [] erato;
  return 0;
}
